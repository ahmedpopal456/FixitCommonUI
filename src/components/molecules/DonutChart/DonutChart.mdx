---
name: DonutChart
menu: 2 - Molecules
---
import { Props, Playground } from "docz";
import { DonutChart } from "./DonutChart.tsx";
import { DoczContainer } from '../../../utils/DoczContainer.tsx';

# DonutChart
A description of the component

# Usage
```tsx
import { DonutChart } from "fixit-common-ui";

<DonutChart value={30} />
```

## Props
<Props of={DonutChart}></Props>

## Default
The Donut chart requires a value to be provided. This value will indicate the percentage
to display.
<Playground>
  <DoczContainer>
    <DonutChart value={25} />
    <DonutChart value={50} />
    <DonutChart value={75} />
    <DonutChart value={100} />
  </DoczContainer>
</Playground>

## Max
If the max value is not provided, the default will be 100. If one is provided,
The value given over the max will form the percentage displayed. If the value is greater
than the max, it will display 100%.
<Playground>
  <DoczContainer>
    <DonutChart value={250} max={1000} />
    <DonutChart value={250} />
  </DoczContainer>
</Playground>

## Color and TextColor
Modify the default color and text color for the Donut Chart.
<Playground>
  <DoczContainer>
    <DonutChart value={75} color="primary" textColor="white" />
    <DonutChart value={75} color="red" textColor="red" />
    <DonutChart value={75} color="green" />
  </DoczContainer>
</Playground>

## Radius
Modify the default color and text color for the Donut Chart.
<Playground>
  <DoczContainer>
    <DonutChart value={75} radius={40} />
    <DonutChart value={75} radius={50} />
    <DonutChart value={75} radius={60} />
    <DonutChart value={75} radius={70} />
    <DonutChart value={75} radius={80} />
  </DoczContainer>
</Playground>

## Stroke Width
Modify the default color and text color for the Donut Chart.
<Playground>
  <DoczContainer>
    <DonutChart value={75} strokeWidth={10} />
    <DonutChart value={75} strokeWidth={30} />
  </DoczContainer>
</Playground>
