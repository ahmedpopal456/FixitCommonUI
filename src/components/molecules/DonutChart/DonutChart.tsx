import React from 'react';
import { View } from 'react-native';

import Svg, { Circle, G } from 'react-native-svg';
import { DonutChartProps } from './DonutChartModel';
import { Label } from '../../atoms/Typography';
import colors from '../../../theme/colors';

/**
 * The DonutChart component is used to display a progress
 * @param donutChartProps - The DonutChart's props
 * @returns A react native custom DonutChart component
 */
export const DonutChart: React.FC<DonutChartProps> = (donutChartProps: DonutChartProps):
JSX.Element => {
  const donutOuterRadius: number = donutChartProps.radius! + donutChartProps.strokeWidth!;
  const circumference: number = 2 * Math.PI * donutChartProps.radius!;

  const percentage = 100 * (donutChartProps.value / donutChartProps.max!);
  const strokeDashoffset = percentage > 100 ? 0 : circumference
  - (circumference * percentage) / 100;

  return (
    <View
      style={{
        width: donutChartProps.radius! * 2,
        height: donutChartProps.radius! * 2,
        alignItems: 'center',
        justifyContent: 'center',
      }}
    >
      <View style={{
        flexDirection: 'row',
        alignSelf: 'center',
        zIndex: 99,
      }}>
        <Label
          fontSize={donutChartProps.radius! / 2}
          color={donutChartProps.textColor}
          fontWeight={500}
          padding={0}
        >
          {percentage > 100 ? '100%' : `${percentage}%`}
        </Label>

      </View>
      <Svg
        height={donutChartProps.radius! * 2}
        width={donutChartProps.radius! * 2}
        viewBox={`0 0 ${donutOuterRadius * 2} ${donutOuterRadius * 2}`}
        style={{ position: 'absolute' }}
      >
        <G rotation="-90" originX={donutOuterRadius} originY={donutOuterRadius}>
          <Circle
            cx="50%"
            cy="50%"
            r={donutChartProps.radius}
            stroke={colors[donutChartProps.color!]}
            strokeWidth={donutChartProps.strokeWidth}
            fill={colors[donutChartProps.color!]}
            fillOpacity={0.2}
            strokeDasharray={circumference}
            strokeDashoffset={strokeDashoffset * -1}
            strokeLinecap="round"
          />
        </G>
      </Svg>
    </View>
  );
};

/** Setting the defaults for the props */
DonutChart.defaultProps = {
  color: 'accent',
  textColor: 'primary',
  radius: 80,
  strokeWidth: 15,
  max: 100,
};
