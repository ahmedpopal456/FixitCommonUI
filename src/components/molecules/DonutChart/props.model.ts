import colors from "../../../theme/colors";
  
  export interface Props {
    /** Can be any color from the theme color palette */
    color?: keyof typeof colors;
    textColor?: keyof typeof colors;
    radius?: number;
    strokeWidth?: number;
    max?: number;
    value: number;
    testID?: string;
  }