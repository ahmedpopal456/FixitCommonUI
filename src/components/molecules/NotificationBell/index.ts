import { NotificationBell } from "./NotificationBell";
export { NotificationBell };
