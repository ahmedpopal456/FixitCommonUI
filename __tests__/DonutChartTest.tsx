import "react-native";
import React from "react";
import { DonutChart } from "../src/components/molecules/DonutChart";

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
import { render } from '@testing-library/react-native';

it("percentage > 100 = 100", () => {
  const {getByText} = render(<DonutChart value={250} max={100} />);
  const percentageDisplayed = getByText("100%");
  expect(percentageDisplayed).toBeTruthy();
});

/** Testing Rendered Output (snapshot testing) */
it("matches default snapshot", () => {
  const donutChart = renderer.create(<DonutChart value={10} />);  
  expect(donutChart.toJSON()).toMatchSnapshot();
});

it("matches color snapshot", () => {
  const donutChart = renderer.create(
    <DonutChart 
      value={10}
      color="accent"
    />);
  
  expect(donutChart.toJSON()).toMatchSnapshot();
});

it("matches textColor snapshot", () => {
  const donutChart = renderer.create(
    <DonutChart 
      value={10}
      textColor="red"
    />);
  
  expect(donutChart.toJSON()).toMatchSnapshot();
});

it("matches radius snapshot", () => {
  const donutChart = renderer.create(
    <DonutChart 
      value={10}
      radius={100}
    />);
  
  expect(donutChart.toJSON()).toMatchSnapshot();
});

it("matches strokeWidth snapshot", () => {
  const donutChart = renderer.create(
    <DonutChart 
      value={10}
      strokeWidth={10}
    />);
  
  expect(donutChart.toJSON()).toMatchSnapshot();
});

it("matches max snapshot", () => {
  const donutChart = renderer.create(
    <DonutChart 
      value={10}
      max={2000}
    />);
  
  expect(donutChart.toJSON()).toMatchSnapshot();
});