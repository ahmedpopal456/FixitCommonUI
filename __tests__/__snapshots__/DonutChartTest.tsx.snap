// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches color snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 160,
      "justifyContent": "center",
      "width": 160,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "zIndex": 99,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        color="primary"
        fontSize={40}
        fontWeight={500}
        justification="left"
        padding={0}
        style={
          Array [
            Object {
              "color": "#33343E",
              "flexDirection": "row",
              "fontFamily": "Raleway",
              "fontSize": 40,
              "fontWeight": "500",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
          ]
        }
      >
        10
        %
      </Text>
    </View>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={160}
    bbWidth={160}
    focusable={false}
    height={160}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "position": "absolute",
        },
        Object {
          "flex": 0,
          "height": 160,
          "width": 160,
        },
      ]
    }
    vbHeight={190}
    vbWidth={190}
    width={160}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            190,
          ]
        }
      >
        <RNSVGCircle
          cx="50%"
          cy="50%"
          fill={4294955338}
          fillOpacity={0.2}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={80}
          stroke={4294955338}
          strokeDasharray={
            Array [
              502.6548245743669,
              502.6548245743669,
            ]
          }
          strokeDashoffset={-452.3893421169302}
          strokeLinecap={1}
          strokeWidth={15}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`matches default snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 160,
      "justifyContent": "center",
      "width": 160,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "zIndex": 99,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        color="primary"
        fontSize={40}
        fontWeight={500}
        justification="left"
        padding={0}
        style={
          Array [
            Object {
              "color": "#33343E",
              "flexDirection": "row",
              "fontFamily": "Raleway",
              "fontSize": 40,
              "fontWeight": "500",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
          ]
        }
      >
        10
        %
      </Text>
    </View>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={160}
    bbWidth={160}
    focusable={false}
    height={160}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "position": "absolute",
        },
        Object {
          "flex": 0,
          "height": 160,
          "width": 160,
        },
      ]
    }
    vbHeight={190}
    vbWidth={190}
    width={160}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            190,
          ]
        }
      >
        <RNSVGCircle
          cx="50%"
          cy="50%"
          fill={4294955338}
          fillOpacity={0.2}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={80}
          stroke={4294955338}
          strokeDasharray={
            Array [
              502.6548245743669,
              502.6548245743669,
            ]
          }
          strokeDashoffset={-452.3893421169302}
          strokeLinecap={1}
          strokeWidth={15}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`matches max snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 160,
      "justifyContent": "center",
      "width": 160,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "zIndex": 99,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        color="primary"
        fontSize={40}
        fontWeight={500}
        justification="left"
        padding={0}
        style={
          Array [
            Object {
              "color": "#33343E",
              "flexDirection": "row",
              "fontFamily": "Raleway",
              "fontSize": 40,
              "fontWeight": "500",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
          ]
        }
      >
        0.5
        %
      </Text>
    </View>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={160}
    bbWidth={160}
    focusable={false}
    height={160}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "position": "absolute",
        },
        Object {
          "flex": 0,
          "height": 160,
          "width": 160,
        },
      ]
    }
    vbHeight={190}
    vbWidth={190}
    width={160}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            190,
          ]
        }
      >
        <RNSVGCircle
          cx="50%"
          cy="50%"
          fill={4294955338}
          fillOpacity={0.2}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={80}
          stroke={4294955338}
          strokeDasharray={
            Array [
              502.6548245743669,
              502.6548245743669,
            ]
          }
          strokeDashoffset={-500.14155045149505}
          strokeLinecap={1}
          strokeWidth={15}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`matches radius snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 200,
      "justifyContent": "center",
      "width": 200,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "zIndex": 99,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        color="primary"
        fontSize={50}
        fontWeight={500}
        justification="left"
        padding={0}
        style={
          Array [
            Object {
              "color": "#33343E",
              "flexDirection": "row",
              "fontFamily": "Raleway",
              "fontSize": 50,
              "fontWeight": "500",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
          ]
        }
      >
        10
        %
      </Text>
    </View>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={200}
    bbWidth={200}
    focusable={false}
    height={200}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "position": "absolute",
        },
        Object {
          "flex": 0,
          "height": 200,
          "width": 200,
        },
      ]
    }
    vbHeight={230}
    vbWidth={230}
    width={200}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            230,
          ]
        }
      >
        <RNSVGCircle
          cx="50%"
          cy="50%"
          fill={4294955338}
          fillOpacity={0.2}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={100}
          stroke={4294955338}
          strokeDasharray={
            Array [
              628.3185307179587,
              628.3185307179587,
            ]
          }
          strokeDashoffset={-565.4866776461628}
          strokeLinecap={1}
          strokeWidth={15}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`matches strokeWidth snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 160,
      "justifyContent": "center",
      "width": 160,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "zIndex": 99,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        color="primary"
        fontSize={40}
        fontWeight={500}
        justification="left"
        padding={0}
        style={
          Array [
            Object {
              "color": "#33343E",
              "flexDirection": "row",
              "fontFamily": "Raleway",
              "fontSize": 40,
              "fontWeight": "500",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
          ]
        }
      >
        10
        %
      </Text>
    </View>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={160}
    bbWidth={160}
    focusable={false}
    height={160}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "position": "absolute",
        },
        Object {
          "flex": 0,
          "height": 160,
          "width": 160,
        },
      ]
    }
    vbHeight={180}
    vbWidth={180}
    width={160}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            180,
          ]
        }
      >
        <RNSVGCircle
          cx="50%"
          cy="50%"
          fill={4294955338}
          fillOpacity={0.2}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={80}
          stroke={4294955338}
          strokeDasharray={
            Array [
              502.6548245743669,
              502.6548245743669,
            ]
          }
          strokeDashoffset={-452.3893421169302}
          strokeLinecap={1}
          strokeWidth={10}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`matches textColor snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 160,
      "justifyContent": "center",
      "width": 160,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "zIndex": 99,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        color="red"
        fontSize={40}
        fontWeight={500}
        justification="left"
        padding={0}
        style={
          Array [
            Object {
              "color": "#D4675A",
              "flexDirection": "row",
              "fontFamily": "Raleway",
              "fontSize": 40,
              "fontWeight": "500",
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "textAlign": "left",
            },
          ]
        }
      >
        10
        %
      </Text>
    </View>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={160}
    bbWidth={160}
    focusable={false}
    height={160}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "position": "absolute",
        },
        Object {
          "flex": 0,
          "height": 160,
          "width": 160,
        },
      ]
    }
    vbHeight={190}
    vbWidth={190}
    width={160}
  >
    <RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            190,
          ]
        }
      >
        <RNSVGCircle
          cx="50%"
          cy="50%"
          fill={4294955338}
          fillOpacity={0.2}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={80}
          stroke={4294955338}
          strokeDasharray={
            Array [
              502.6548245743669,
              502.6548245743669,
            ]
          }
          strokeDashoffset={-452.3893421169302}
          strokeLinecap={1}
          strokeWidth={15}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
